code/modules/atmospherics/gasmixtures/reactions.dm

```diff
@@ -291,6 +291,60 @@ datum/gas_reaction/freonfire/react(datum/gas_mixture/air, datum/holder)
 		if(new_heat_capacity > MINIMUM_HEAT_CAPACITY)
 			air.temperature = (temperature*old_heat_capacity + energy_released)/new_heat_capacity
 
+/datum/gas_reaction/h2fire
+	priority = -1 //fire should ALWAYS be last, but tritium fires happen before plasma fires
+	name = "Hydrogen Combustion"
+	id = "h2fire"
+
+/datum/gas_reaction/h2fire/init_reqs()
+	min_requirements = list(
+		"TEMP" = FIRE_MINIMUM_TEMPERATURE_TO_EXIST,
+		/datum/gas/hydrogen = MINIMUM_MOLE_COUNT,
+		/datum/gas/oxygen = MINIMUM_MOLE_COUNT
+	)
+
+/datum/gas_reaction/h2fire/react(datum/gas_mixture/air, datum/holder)
+	var/energy_released = 0
+	var/old_heat_capacity = air.heat_capacity()
+	var/list/cached_gases = air.gases //this speeds things up because accessing datum vars is slow
+	var/temperature = air.temperature
+	var/list/cached_results = air.reaction_results
+	cached_results["fire"] = 0
+	var/turf/open/location = isturf(holder) ? holder : null
+	var/burned_fuel = 0
+	if(cached_gases[/datum/gas/oxygen][MOLES] < cached_gases[/datum/gas/hydrogen][MOLES] || MINIMUM_TRIT_OXYBURN_ENERGY > air.thermal_energy())
+		burned_fuel = cached_gases[/datum/gas/oxygen][MOLES]/TRITIUM_BURN_OXY_FACTOR
+		cached_gases[/datum/gas/hydrogen][MOLES] -= burned_fuel
+	else
+		burned_fuel = cached_gases[/datum/gas/hydrogen][MOLES]*TRITIUM_BURN_TRIT_FACTOR
+		cached_gases[/datum/gas/hydrogen][MOLES] -= cached_gases[/datum/gas/hydrogen][MOLES]/TRITIUM_BURN_TRIT_FACTOR
+		cached_gases[/datum/gas/oxygen][MOLES] -= cached_gases[/datum/gas/hydrogen][MOLES]
+
+	if(burned_fuel)
+		energy_released += (FIRE_HYDROGEN_ENERGY_RELEASED * burned_fuel)
+
+		ASSERT_GAS(/datum/gas/water_vapor, air) //oxygen+more-or-less hydrogen=H2O
+		cached_gases[/datum/gas/water_vapor][MOLES] += burned_fuel/TRITIUM_BURN_OXY_FACTOR
```

Comment:
You should probably make your own version of these defines, just so it's more clear that this exists.

Reply:
done next commit